Add propagating overloaded from replicas

psarna opened the pull request:
This series adds propagating overloaded exceptions from replicas to the coordinator, which in turn allows propagating them to the users. The motivation behind it is that drivers are capable of applying retry policies for specific kinds of errors, so it's better to provide as much context as possible.

The way of propagating more context is via a new `query::status` field returned from read operations. The field contains additional information what happened - currently, the status can be either `OK` or `OVERLOADED`. `OK` simply means that the query succeeded, while `OVERLOADED` hints that the result should not be returned to the user at all, but instead discarded and translated to `OverloadedException`. That way, when a replica gets overloaded, Scylla will be able to return a more meaningful error type than a generic read failure.

Note that in C* world, `OverloadedException` is currently understood as `this coordinator got overloaded`, and this series makes this meaning more generic, since it's a replica (or multiple replicas) that failed. Retry policies picked by the drivers should consider that distinction, since some policies may make less sense if the overloaded exception was actually propagated from a replica. One of such policies is "try on next coordinator", which is not guaranteed to help if it's a remote replica that got overloaded. Still, other policies (e.g. reduce the load) may be very effective.

gleb-cloudius commented on database.cc:
```diff
@@ -1426,7 +1426,14 @@ database::query(schema_ptr s, const query::read_command& cmd, query::result_opti
+            try {
+                f.get();
+            } catch (exceptions::overloaded_exception& e) {
+                return make_ready_future<std::tuple<lw_shared_ptr<query::result>, cache_temperature, query::status>>(std::tuple(make_lw_shared<query::result>(), hit_rate, query::status::OVERLOADED));
```
Note this is not backwards compatible. If an old node will get this it will consider it as empty value not as failed read.

psarna replied:
Oops, right. I guess that piggy-backing on an existing tuple won't work for upgrades then, and a new verb would be a way to go...

psarna replied:
Or, I can just introuduce a cluster feature.

gleb-cloudius replied:
Different verb will require cluster feature anyway.

gleb-cloudius commented:
I see that you covered all relevant verbs: read data, mutation data and digest so compiler should handle all the rest and make sure that you do not miss any places where the data is produces. One question though: wouldn't is be better to handle errors at the highest level in storage_proxy just before returning the values?

avikivity commented on query-status.hh:
```diff
@@ -0,0 +1,47 @@
+
+class status {
+public:
+    enum value { OK, OVERLOADED };
```
Why not `enum class status` instead of `class status`?

psarna replied:
Can enum class have methods? I need something to allow serialization, so I went with a regular class with a method which translates the value to uint8_t. I always avoided enum classes because they ended up being painful instead of easier to use.

denesb replied:
I thought we already have an enum serializer. Or maybe I just wanted to write one and didn't end up using it. Anyway, it is trivial to write one.

psarna commented:
> I see that you covered all relevant verbs: read data, mutation data and digest so compiler should handle all the rest and make sure that you do not miss any places where the data is produces. One question though: wouldn't is be better to handle errors at the highest level in storage_proxy just before returning the values?

Sure, but where would that be? I tried to pick a good place, but it's easy to get lost in here, since futures are often waited-for indirectly.

gleb-cloudius commented:
On Wed, Dec 09, 2020 at 04:20:42AM -0800, Piotr Sarna wrote:
> > I see that you covered all relevant verbs: read data, mutation data and digest so compiler should handle all the rest and make sure that you do not miss any places where the data is produces. One question though: wouldn't is be better to handle errors at the highest level in storage_proxy just before returning the values?
> 
> Sure, but where would that be? I tried to pick a good place, but it's easy to get lost in here, since futures are often waited-for indirectly.
> 
What I had in mind is in lambdas that are passed to register_read_data, register_read_mutation_data and register_read_digest.

--
			Gleb.

psarna commented:
Ooooh you mean the replica side. Now that makes sense, yup, I'll move everything to storage proxy to keep database.cc intact.

psarna commented:
Turned out to be a great idea - it can be neatly handled with `handle_exception_type()` in storage proxy. I'll push that in v2, along with a cluster feature for backward compatibility.

psarna commented:
v2:
 * added a cluster feature for backward compatibility
 * made returning the overloaded status conditional on the aforementioned cluster feature - if the cluster does not support the feature, a regular error will be returned as before
 * moved checking the status one level up to storage proxy - to the lambdas where handlers are registered

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -2763,18 +2763,21 @@ class abstract_read_resolver {
     }
     virtual ~abstract_read_resolver() {};
-    virtual void on_error(gms::inet_address ep, bool disconnect) = 0;
+    virtual void on_error(gms::inet_address ep, bool disconnect, std::exception_ptr overloaded) = 0;
```
Instead of passing a pointer to an exception here, why not change `bool disconnect` to `error_type type` instead where error_type can be `enum {disconnect, overloaded}`

psarna replied:
I used the exception pointer since I already had it from catching the exception, but an enum also makes sense. I'll soon push an unrelated PR which adds an error enum to storage proxy anyway, so I'll wait for that one to be reviewed first. If it goes in, I can just rebase and use the already existing enum.

gleb-cloudius commented:
On Thu, Dec 10, 2020 at 03:14:20AM -0800, Botond Dénes wrote:
> @denesb commented on this pull request.
> 
> 
> 
> > + * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
> + */
> +
> +#pragma once
> +#include <cinttypes>
> +
> +namespace ser {
> +template <typename T>
> +class serializer;
> +}
> +
> +namespace query {
> +
> +class status {
> +public:
> +    enum value { OK, OVERLOADED };
> 
> I thought we already have an enum serializer. Or maybe I just wanted to write one and didn't end up using it. Anyway, it is trivial to write one.
> 
Having it as a structure will allow us to add more into into the status
later on if we wan to.

--
			Gleb.

psarna commented:
v2:
 * switched to using an error enum instead of a boolean + an exception pointer for handling different error types

nyh commented on idl/status.idl.hh:
```diff
@@ -0,0 +1,28 @@
+
+namespace query {
+
+class status {
```
I think we need a comment what this means.

nyh commented on gms/feature.hh:
```diff
@@ -145,6 +145,7 @@ extern const std::string_view PER_TABLE_CACHING;
 extern const std::string_view DIGEST_FOR_NULL_VALUES;
 extern const std::string_view CORRECT_IDX_TOKEN_IN_SECONDARY_INDEX;
 extern const std::string_view ALTERNATOR_STREAMS;
+extern const std::string_view QUERY_STATUS_FOR_READS;
```
Commenting here because I can't comment on the entire patch :-(
I didn't 'understand why a flag is needed. When we add a new optional part at the end of RPC requests or replies, we already have a mechanism to support a mixture of old and new nodes - an old node will ignore the "status" field sent by a new node, and a new node will know it didn't get a status from an old node.  So why do we need the extra cluster-wide flag? It would be nice to explain this in the commit message.

psarna replied:
https://github.com/scylladb/scylla/pull/7773#discussion_r539194347

nyh replied:
It looks like you have two different mechanisms for indicating failure - the pre-existing error, and the new "status". Why do we need both mechanisms?
Also, even if you do have both, if an error would produce *both* an error and a status, even an old node would understand that this is a failure. No?

psarna replied:
There's no way of passing an error type via Seastar rpc, and following the discussion here https://github.com/scylladb/seastar/issues/835 , I went with propagating the error type as a field. If I get it correctly, the old error is generally supposed to be an error related to RPC, as in "calling the remote procedure failed", while the status field has application logic - e.g. that a read failed because a replica got overloaded.
I'm pretty sure that the RPC layer allows sending either data or an error, but not both. If I'm not correct, then your suggestion is applicable. @gleb-cloudius?

gleb-cloudius replied:
I have no idea what @nyh has in mind. If he have a concrete proposal he need to expend on it. Obviously if you want to return something else instead of an error you need to make sure that the receiving side can interpret this something correctly.

nyh replied:
My thinking was that just like in Seastar both the success and the failure have types (i.e., the exception contains an object, not just the fact that an exception happened), so should RPC have a way to say which error happened, and I was. Without this feature, now when Piotr wants to return a special type of error, he is forced to return a success (!) plus an extra result field which says error, but as we noticed the result is a mess - we can't use the new mechanism if the entire cluster doesn't support it (or old nodes will think the reply is a success, while it's meant to be a type of error), and it is undefined what exactly happens if the result contains both the actual result and an error code.

I realize we can't overhaul RPC now, just for this feature. However, I raised to Piotr one idea that could allow the read verb to return multiple types of errors without changing RPC. RPC errors already have string messaes. We could (ab)use the string message in the same way that HTTP does this: begin the string with a numeric error code. This can easily be made backward-compatible, because today - I assume - the string the existing code returns does not begin with a numeric error code, so if new nodes start to return numeric error codes then 1. old nodes will still consider those responses as errors, and 2. new nodes receving a response from an old node will know this is an error.

gleb-cloudius replied:
Piotr is not forced to return a success. He can introduce another verb that returns variant<result, error> but that requires introducing another verb. This and why it is not RPC job to transport user exception is discussed in https://github.com/scylladb/seastar/issues/835. Did you read it?

I cannot really consider using error string as a way to transport error codes seriously. It is fine for hacking a quick prototype, but not a proper protocol.

avikivity replied:
Adding a new verb isn't a problem. It requires hassle (cluster feature to know if we can call it), but it's better than a hack that we'll have to work around forever afterwards.

gleb-cloudius replied:
Which hack are you talking about?

nyh replied:
It's a "hack" to have a verb return both the success and error code, and hope nobody returns **both** of them because we don't know what that is supposed to mean.

nyh replied:
> I cannot really consider using error string as a way to transport error codes seriously. It is fine for hacking a quick prototype, but not a proper protocol.

Just out of curiosity, what did you imagine to be the use of these error strings, if we cannot use them to tell error types apart? If you don't like error codes as *part* of the string, how about returning different error strings (no parsing - just comparing the full string) - or is that also not a proper protocol.

gleb-cloudius replied:
The verb does not return success. Where did you get that?  It is definitely not as clean as introducing another verb with variant<> return value, but it is much less hassle and I do not see why would we need to "work around it forever afterwards".

gleb-cloudius replied:
> > I cannot really consider using error string as a way to transport error codes seriously. It is fine for hacking a quick prototype, but not a proper protocol.
> 
> Just out of curiosity, what did you imagine to be the use of these error strings, if we cannot use them to tell error types apart?
Logged for information purpose.

> If you don't like error codes as _part_ of the string, how about returning different error strings (no parsing - just comparing the full string) - or is that also not a proper protocol.

Of course not.

nyh commented on exceptions/exceptions.hh:
```diff
@@ -228,6 +228,7 @@ struct read_failure_exception : public request_failure_exception {
 
 struct overloaded_exception : public cassandra_exception {
+    overloaded_exception() noexcept : cassandra_exception(exception_code::OVERLOADED, "") {};
     explicit overloaded_exception(size_t c) noexcept :
```
Not related to this patch, but since I'm looking at this code now, I want to comment that if we start using overloaded_exception more, we should get rid of this hints-specific constructor, and move this message to the hints code which uses it.

nyh commented on service/storage_proxy.hh:
```diff
@@ -158,6 +158,8 @@ public:
         TIMEOUT,
         FAILURE,
+        DISCONNECTED,
+        OVERLOADED,
```
I really think we need to explain the meaning of all these errors somewhere. What does "DISCONNECTED" mean, disconnected from what?

nyh commented on service/storage_proxy.cc:
```diff
@@ -4959,7 +4988,14 @@ void storage_proxy::init_messaging_service() {
+                    if (p->features().cluster_supports_query_status_for_reads()) {
+                        return make_ready_future<rpc::tuple<foreign_ptr<lw_shared_ptr<query::result>>, cache_temperature, query::status>>(
+                                rpc::tuple(make_foreign(make_lw_shared<query::result>()), cache_temperature::invalid(), query::status::OVERLOADED));
+                    } else {
```
I still don't understand why you can't do this - return "status" unconditionally - without checking any flag. Isn't this the whole point of the IDL, that we can add new elements in structures and old nodes will just ignore them?

psarna replied:
https://github.com/scylladb/scylla/pull/7773#discussion_r539194347

nyh commented on database.cc:
```diff
@@ -1407,7 +1407,7 @@ compare_atomic_cell_for_merge(atomic_cell_view left, atomic_cell_view right) {
 }
 
-future<std::tuple<lw_shared_ptr<query::result>, cache_temperature>>
+future<std::tuple<lw_shared_ptr<query::result>, cache_temperature, query::status>>
```
Ok, after some more discussions about RPC, I understand now that the intention was that these functions should return *either* a result *or* a non-OK status. I think this needs to be clearly documented, it's really not obvious.

But @psarna I think I have an even better (maybe...) idea:
You can leave this function's signature unchanged, and translate a non-OK status into an exceptional future return. You need this weird "status" thing in the RPC function, but that is the *only* place where you need it! In all the layers you now added "status" too, you could have left the unchanged if the lowest later translates the non-OK status into an expection

psarna replied:
Adding `query::status` here was not done to propagate exceptions, it's actually just a convenience change to avoid adding another continuation to provide a `OK` status. With this function returning a status, the RPC handler looks like that (pseudocode):
```cpp
return query_result_local(stuff).handle_exception_type([] (overloaded_exception& e {
    return ready_future_with_overloaded_exception;
})
```
, while if I leave this function's declaration as it was before, I'd have to do:
```cpp
return query_result_local(stuff).then([] (std::tuple<query_result, cache_temperature>) {
    return make_ready_future(query_result, cache_temperature, status::OK);
}).handle_exception_type([] (overloaded_exception& e {
    return ready_future_with_overloaded_exception;
})
```

Still, I agree there is value in keeping the interfaces clean, so I'll try removing status from non-storage-proxy functions and see if it looks alright.

psarna commented:
One more thing to consider: currently, the Cassandra world seems to treat OverloadedException as "coordinator got overloaded", which isn't really in line with "replica got overloaded".
Proof from CQL v4 spec:
```
    0x1001    Overloaded: the request cannot be processed because the
              coordinator node is overloaded
```
The potential issue with this assumption is that some retry policies applied by the drivers may make sense for coordinators being overloaded, and be useless for overloaded replicas (e.g. "try on another coordinator retry policy").

It may still be useful to let the client know that something got overloaded, because if the drivers' policy would be to back off and slow down the number of requests, then it still makes a lot of sense for overloaded replicas as well. Still, it's something we should keep in mind when propagating OverloadedException.

psarna commented:
v3:
 * removed `query::status` from most of the functions, except the ones in storage proxy responsible for RPC communication. That makes majority of internal interfaces free of the additional field at the cost of a single continuation in RPC handling.
 * added comments to errors
 * added a comment blurb to the query status class
 * the PR description, which will act as commit message, will be updated in a minute as well

Not done (yet): I didn't go with the new verb approach, because I'm not entirely sure it will be cleaner - it would result in duplicating the code for each verb, and there are three of them, while the `status` field could possibly be useful later anyway, to provide more information from replicas (e.g. *how* overloaded they feel right now)

avikivity commented:
Why would a new verb cause duplication? all verbs would call a common implementation.

psarna commented:
Right, most of the code could be extracted to a common core, I meant the duplication of methods in storage proxy (maybe explosion would be a better wording) - so that instead of
 * register_read_data
 * register_read_mutation_data
 * register_read_digest

we get 
* register_read_data
* register_read_mutation_data
* register_read_digest
* register_read_data_or_error
* register_read_mutation_data_or_error
* register_read_digest_or_error

(Not to mention that I don't have a clear idea how to name the new API... )

I can try though, maybe it looks more painful in my imagination than it really is.

avikivity commented:
I'd call the new verb read_data() and the old verb old_read_data().

I suppose you can make register_read_data() accept two callbacks, one old and one new.

btw (@gleb-cloudius) the fact that there is callback code in init_messaging_service() is confusing, since it makes the stack traces point to init_messaging_service::lambda rather than some named function that indicates the verb called.

gleb-cloudius commented:
On Wed, Dec 16, 2020 at 06:57:18AM -0800, Avi Kivity wrote:
> I'd call the new verb read_data() and the old verb old_read_data().
> 
> I suppose you can make register_read_data() accept two callbacks, one old and one new.
> 
Please no. I want this register_() function to be generated from idl
eventually.

Before we are moving to a new verb did you consider Piotr's idea where
the status can be changed to non binary thing, so the data can be
returned along with a status that says how much a replica is overloaded,
so that a coordinator may consider choosing other replicas.

> btw (@gleb-cloudius) the fact that there is callback code in init_messaging_service() is confusing, since it makes the stack traces point to init_messaging_service::lambda rather than some named function that indicates the verb called.
> 
OTOH having init_messaging_service there points to a fact that the
problem is in an RPC handler where from a function name it may be not
obvious :) Named function can be passed there as well of course, passing
a lambda is just a convenience. There are a lot of traces where I wish
lambdas would have been a named functions!

--
			Gleb.

avikivity commented:
I missed the comment, but returning a variant<good result, one kind of error result with payload, another kind of error result with another payload> sounds like a good idea, if that's what is meant.

avikivity commented:
> On Wed, Dec 16, 2020 at 06:57:18AM -0800, Avi Kivity wrote: I'd call the new verb read_data() and the old verb old_read_data(). I suppose you can make register_read_data() accept two callbacks, one old and one new.
> Please no. I want this register_() function to be generated from idl eventually. Before we are moving to a new verb did you 

We could teach IDL that one function obsoletes another, and so to generate a single registration entry for them.

psarna commented:
> I missed the comment, but returning a variant<good result, one kind of error result with payload, another kind of error result with another payload> sounds like a good idea, if that's what is meant.

The comment was actually an attempt to defend the current implementation of adding a status field to the existing tuple. That way, some statuses (like OVERLOADED) will imply that no meaningful result was returned, but other ones (which are not here yet) could coexist with a valid response.

gleb-cloudius commented:
On Wed, Dec 16, 2020 at 07:16:49AM -0800, Avi Kivity wrote:
> I missed the comment, but returning a variant<good result, one kind of error result with payload, another kind of error result with another payload> sounds like a good idea, if that's what is meant.
> 
It is more of a tuple<good result, overload level> where overload level
can be on max and then the result should not be used. With variant it
can be written as variant<tuple<good result, overload level>, error>,
but if we are adding overload level anyway going through the hassle of
moving to another verb look less attractive IMO.

--
			Gleb.

avikivity commented:
How can you add more info without a new verb?

psarna commented:
By appending optionals to a tuple, just like with `query::status` in this series. E.g. `rpc::tuple<result, cache_temperature, query::status, query::status::payload>`

gleb-cloudius commented:
On Wed, Dec 16, 2020 at 07:31:55AM -0800, Avi Kivity wrote:
> How can you add more info without a new verb?
> 
Sometimes our rpc can do magic!

--
			Gleb.

psarna commented:
I was working on an example of using a new verb with std::variant, but serializers keep screaming at me that they cannot find a proper instantiation for a tuple inside a variant... so I haven't succeeded today. I'll try to provide a sample soonish.

psarna commented:
I'm closing this pull request, because I changed my mind and I think it's based on false assumptions - `OverloadedException` really *does* mean that a coordinator node is overloaded, so returning this status for overloaded replicas will only confuse the retry policies in the drivers.
